rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,1625 @@
+   pipeline is specifically for system events, but Heapster is deprecated?
+   This leads me to my next question...~~
+   Heapster is indeed outdated, as well as Honeycomb integration guide.
+   Kubernetes now solves it's internal autoscaling pipelines needs with
+   [`metrics-server`][metrics-server] - a similar idea yet much more lightweight
+   implementation.
+1. ~~How are we collecting Kubernetes system events? Is that outside of the
+   scope of this RFC? And why does this take an entirely different path?
+   (ref [issue#1293])~~
+   See the [Other data gathering][anchor_other_data_gathering] section.
+1. What are some of the details that set Vector's Kubernetes integration apart?
+   This is for marketing purposes and also helps us "raise the bar".
+
+### From Mike
+
+1. What significantly different k8s cluster "flavors" are there? Which ones do
+   we want to test against? Some clusters use `docker`, some use `CRI-O`,
+   [etc][container_runtimes]. Some even use [gVisor] or [Firecracker]. There
+   might be differences in how different container runtimes handle logs.
+1. How do we want to approach Helm Chart Repository management.
+1. How do we implement liveness, readiness and startup probes?
+   Readiness probe is a tricky one. See [Container probes](#container-probes).
+1. Can we populate file at `terminationMessagePath` with some meaningful
+   information when we exit or crash?
+1. Can we allow passing arbitrary fields from the `Pod` object to the event?
+   Currently we only to pass `pod_id`, pod `annotations` and pod `labels`.
+
+## Plan Of Attack
+
+- [ ] Agree on minimal Kubernetes version.
+- [ ] Agree on a list of Kubernetes cluster flavors we want to test against.
+- [ ] Setup a proper testing suite for k8s.
+  - [ ] Support for customizable k8s clusters. See [issue#2170].
+  - [ ] Look into [issue#2225] and see if we can include it as part of this
+        work.
+  - [ ] Stabilize k8s integration tests. See [issue#2193], [issue#2216],
+        and [issue#1635].
+  - [ ] Ensure we are testing all supported minor versions. See
+        [issue#2223].
+- [ ] Audit and improve the `kubernetes` source.
+  - [ ] Handle the log recursion problem where Vector ingests it's own logs.
+        See [issue#2218] and [issue#2171].
+  - [ ] Audit the `file` source strategy. See [issue#2199] and [issue#1910].
+    - [ ] Merge split logs. See [pr#2134].
+- [ ] Audit and improve the `kubernetes_pod_matadata` transform.
+  - [ ] Use the `log_schema.kubernetes_key` setting. See [issue#1867].
+- [ ] Add a way to load optional config files (i.e. load config file if it
+      exists, and ignore it if it doesn't). Required to elegantly load multiple
+      files so that we can split the configuration.
+- [ ] Add `kubernetes` source reference documentation.
+- [ ] Prepare YAML deployment config.
+- [ ] Prepare Heml Chart.
+- [ ] Prepare Heml Chart Repository.
+- [ ] Integrate kubernetes configuration snapshotting into the release process.
```

Comment:
I'd prefer that we move this up above the Helm charts. Correct me if I'm wrong, but I think we can layer on the Helm integration once we are fully done with the Yaml process, right? This reason I want to move the Helm work to the backend is that it's possible we might defer it.

Reply:
While Helm Chart also produces YAML files - it's way more complicated than YAML files. Ideally, we want to it the other way around - finish the Helm Chart first, and then get generate YAML file through it so that there's a single source of truth.
Overall, I expect the preparation of the Chart templates to take way more effort than YAML files since the Helm Charts is where we want to address all the "sane defaults" stuff. Since YAML files are just a single configuration, and Helm Charts usually render to a matrix of those, we'd rather work with Helm Charts first.
I have to add this to the RFC body.

Reply:
Got it. That makes sense me.

Reply:
Added to the RFC.